worlds/jakanddaxter/__init__.py

```diff
@@ -0,0 +1,505 @@
+
+    # Handles various options validation, rules enforcement, and caching of important information.
+    def generate_early(self) -> None:
+
+        # Initialize the level-region dictionary.
+        self.level_to_regions = defaultdict(list)
+        self.level_to_orb_regions = defaultdict(list)
+
+        # Cache the power cell threshold values for quicker reference.
+        self.power_cell_thresholds = [
+            self.options.fire_canyon_cell_count.value,
+            self.options.mountain_pass_cell_count.value,
+            self.options.lava_tube_cell_count.value,
+            100,  # The 100 Power Cell Door.
+        ]
+
+        # Order the thresholds ascending and set the options values to the new order.
+        if self.options.enable_ordered_cell_counts:
+            self.power_cell_thresholds.sort()
+            self.options.fire_canyon_cell_count.value = self.power_cell_thresholds[0]
+            self.options.mountain_pass_cell_count.value = self.power_cell_thresholds[1]
+            self.options.lava_tube_cell_count.value = self.power_cell_thresholds[2]
+
+        # For the fairness of other players in a multiworld game, enforce some friendly limitations on our options,
+        # so we don't cause chaos during seed generation. These friendly limits should **guarantee** a successful gen.
+        # We would have done this earlier, but we needed to sort the power cell thresholds first.
+        enforce_friendly_options = self.settings.enforce_friendly_options
+        if enforce_friendly_options:
+            if self.multiworld.players > 1:
+                from .Rules import enforce_multiplayer_limits
+                enforce_multiplayer_limits(self)
+            else:
+                from .Rules import enforce_singleplayer_limits
+                enforce_singleplayer_limits(self)
+
+        # Calculate the number of power cells needed for full region access, the number being replaced by traps,
+        # and the number of remaining filler.
+        if self.options.jak_completion_condition == Options.CompletionCondition.option_open_100_cell_door:
+            self.total_prog_cells = 100
+        else:
+            self.total_prog_cells = max(self.power_cell_thresholds[:3])
+        non_prog_cells = self.total_power_cells - self.total_prog_cells
+        self.total_trap_cells = min(self.options.filler_power_cells_replaced_with_traps.value, non_prog_cells)
+        self.options.filler_power_cells_replaced_with_traps.value = self.total_trap_cells
+        self.total_filler_cells = non_prog_cells - self.total_trap_cells
+
+        # Verify that we didn't overload the trade amounts with more orbs than exist in the world.
+        # This is easy to do by accident even in a singleplayer world.
+        self.total_trade_orbs = (9 * self.options.citizen_orb_trade_amount) + (6 * self.options.oracle_orb_trade_amount)
+        from .Rules import verify_orb_trade_amounts
```

Comment:
These floating imports are considered bad style

Doesn't matter that much, I'm just confused why they're here

Reply:
Wilco.

Reply:
Update: moving those imports to the top of the file leads to circular imports. I'm pretty sure that's why I put them where they were to begin with - those imports rely on the world class. 

Any advice?

Reply:
Why do they rely on the world class?

Reply:
Oh is it for type checking? You should check some other worlds for how they do that. It's a combination of using str typing `: "JakAndDaxterWorld"` and importing via `if __type_checking__: import JakAndDaxterWorld` or something like that